tests/test_mariadb.py

```diff
@@ -276,3 +280,76 @@ def test_mariadb_healthcheck_galera_cluster_disabled(auto_container_per_test):
     _wait_for_server(conn)
 
     conn.run_expect([1], "healthcheck.sh --su-mysql --galera_online")
+
+
+_DB_ENV = {
+    "MARIADB_USER": _OTHER_DB_USER,
+    "MARIADB_PASSWORD": _OTHER_DB_PW,
+    "MARIADB_DATABASE": _TEST_DB,
+    "MARIADB_ROOT_PASSWORD": MARIADB_ROOT_PASSWORD,
+    "MARIADB_AUTO_UPGRADE": "1",
+}
+
+
+@pytest.mark.parametrize("ctr_image", MARIADB_CONTAINERS)
+@pytest.mark.skipif(
+    OS_VERSION in ("15.5",), reason="MariaDB upgrade scenario not supported"
+)
+def test_mariadb_upgrade(
+    container_runtime: OciRuntimeBase,
+    pytestconfig: pytest.Config,
+    ctr_image: DerivedContainer,
+    tmp_path: Path,
+) -> None:
+    mounts = [BindMount(host_path=tmp_path, container_path="/var/lib/mysql")]
+    mariadb_old = DerivedContainer(
+        base="registry.suse.com/suse/mariadb:10.6",
+        containerfile='RUN set -euo pipefail; head -n -1 /usr/local/bin/gosu > /tmp/gosu;  echo \'exec setpriv --pdeathsig=keep --reuid="$u" --regid="$u" --clear-groups -- "$@"\' >> /tmp/gosu;  mv /tmp/gosu /usr/local/bin/gosu; chmod +x /usr/local/bin/gosu',
+        volume_mounts=mounts,
+        extra_environment_variables=_DB_ENV,
+    )
+    mariadb_new = DerivedContainer(
+        base=ctr_image,
+        volume_mounts=mounts,
+        extra_environment_variables=_DB_ENV,
+    )
+    with ContainerLauncher.from_pytestconfig(
+        mariadb_old, container_runtime, pytestconfig
+    ) as launcher:
+        launcher.launch_container()
+        con = launcher.container_data.connection
+        _wait_for_server(con)
+        mariadb_cmd = f"mariadb --user={_OTHER_DB_USER} --password={_OTHER_DB_PW} --host=0.0.0.0 {_TEST_DB}"
+
+        con.check_output(
+            f'echo "CREATE TABLE test (id serial PRIMARY KEY, num integer, data varchar(32));" | {mariadb_cmd}'
+        )
+        con.check_output(
+            f"echo 'INSERT INTO test (num, data) VALUES (100, \"abcdef\")' | {mariadb_cmd}"
```

Comment:
According to docs:

- Updating the system tables in the mysql database to the newest version. This is very quick.
- [mariadb-upgrade](https://mariadb.com/kb/en/mariadb-upgrade/) also runs [mariadb-check --check-upgrade](https://mariadb.com/kb/en/mariadb-check/) to check if there have been any collation changes between the major versions. This recreates indexes in old tables that are using any of the changed collations. This can take a bit of time if there are a lot of tables or there are many tables which used the changed collation. The last time a collation changed was in MariaDB/MySQL 5.1.23.

It would be nice to touch system tables, such as changing the database charset to a non-standard one.

It would be nice to have a data file that has multiple commands and tables. Ideally creating indexes, so we know that MariaDB can migrate indexes, plus perhaps including more data types than just integer and varchar.

Regular tables are not very likely to break or change unless there's a change in the data type of a column I believe.

Reply:
Do we have any existing SQL file that has a good load of data(we might have used in earlier experiments) or would you like to recommend any specific dataset for this or we can create our own random data(with indexes for sure)?

Reply:
I don't have any, maybe the maintainer has one. Otherwise, we could grab some dataset from Kaggle if needed.

There's no need to be a big dump, just having a few tables, indexes, and changing things that touch system tables would be sufficient I believe.

Reply:
Additional info: 

As per the official mariadb upgrade tests here: https://github.com/MariaDB/server/blob/fd0cc2b1fdee24955d39cdcecbc457324b85b2d2/tests/upgrade_from/test_upgrade.sh, they create a [simple dataset ](https://github.com/MariaDB/server/blob/fd0cc2b1fdee24955d39cdcecbc457324b85b2d2/tests/upgrade_from/test_upgrade.sh#L213)for testing the upgrade.

So, i think we don't have to create dataset specific to each versions and the dataset can be common across all the version. But, as you suggested the dataset can have indices and we can also touch the system tables.

Reply:
We have decided to use the standard data set used in official mariadb test cases.